README.md
# node-red-contrib-iiot-rpi-max7219

[![platform](https://img.shields.io/badge/platform-Node--RED-red)](https://nodered.org)
[![platform](https://img.shields.io/badge/platform-Raspberry--Pi-ff69b4)](https://www.raspberrypi.com/)
![NPM version](https://badge.fury.io/js/node-red-contrib-iiot-rpi-max7219.svg)
![NPM](https://img.shields.io/npm/l/node-red-contrib-iiot-rpi-max7219)
[![Donate](https://img.shields.io/badge/Donate-PayPal-yellow.svg)](https://www.paypal.com/cgi-bin/webscr?cmd=_s-xclick&hosted_button_id=ZDRCZBQFWV3A6)

A Node-Red node for max7219 8-Digit LED Display<br>

Example  **max7219** 7-Segment Module

![image info](images/max7219-7seg.png)

Example  **max7219** Matrix Module

![image info](images/max7219-matrix.png)

## Nodes
- **max7219-7seg**: Controls mx7219 7-Segment Modules.
- **max7219-mtx**: Controls mx7219 Matrix Modules.

## Installation
Install with Node-Red Palette Manager or npm command:
```
cd ~/.node-red
npm install node-red-contrib-iiot-rpi-max7219
```

## Usage

[How To Use](howtouse.md)<br>

## Changelog
[CHANGELOG](CHANGELOG.md)<br>

CHANGELOG.md
1.0.20
- Initial Release

1.0.21
- Add write and display custom font

1.0.22
- Add write and display custom font for 7-segment led
- New HowTo document


lib/syslib.js
/**
 * Copyright 2022 Ocean (iiot2k@gmail.com).
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use node file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 **/

"use strict";

const fs = require('fs');

module.exports.outError = function (node, errShort, errLong) {
	if (node.save_txt === errShort)
		return true;

	node.save_txt = errShort;
	node.status({ fill: "red", shape: "ring", text: errShort });
	node.error(errLong);

	return true;
}

module.exports.outOk = function (node) {
	if (node.save_txt === "-OK-")
		return;

	node.save_txt = "-OK-";
	node.status({ fill: "green", shape: "ring"});
}

function getRpiTarget()
{
	if (process.platform !== 'linux')
		return undefined;
	
	var cpuinfo = fs.readFileSync("/proc/cpuinfo").toString();

	if (cpuinfo.indexOf(": BCM") === -1)
		return undefined;
	
	if (process.arch === "arm64")
		return "_s64.node";
	else if (process.arch === "arm")
		return "_s32.node";
	
	return undefined;
}

module.exports.LoadModule = function (module) {
	try {
		var modName = getRpiTarget();
		if (modName !== undefined)
			return require("./" + module + modName);
	}
	catch (e) { console.log(e); }
	return undefined;
}


max7219-7seg.js
/**
 * Copyright 2022 Ocean (iiot2k@gmail.com).
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 **/

"use strict";

module.exports = function(RED) {
	const syslib = require("./lib/syslib.js");
	const sysmodule = syslib.LoadModule("rpi_max7219");

    RED.nodes.registerType("max7219-7seg", function(n) {
		var node = this;
		RED.nodes.createNode(node, n);

		node.spich = parseInt(n.spich);
		node.nmod = n.nmod;
		node.npos = n.npos;
		node.intens = n.intens;
		node.ralign = n.ralign;
		node.fill = n.fill;
		node.iserror = false;
		node.status({});

		node.name = this.name || "7seg-led " + node.spich + "#" + node.npos + "/" + node.nmod;

		if (sysmodule === undefined)
			node.iserror = syslib.outError(node, "driver error", "driver not load, wrong os or not Raspi");
		else if (node.npos >= n.nmod)
			node.iserror = syslib.outError(node, "inv position", "invalid postion or pos >= number module");
		else if (!sysmodule.init(node.spich, node.nmod, node.intens))
			node.iserror = syslib.outError(node, "open error", "spi port not open, check spi settings");

		node.on("input", function (msg) {
			if (node.iserror)
				return;
			
			if (msg.hasOwnProperty('clear'))
				sysmodule.clear(node.spich);

			if (msg.hasOwnProperty('intens')) {
				if ((typeof msg.intens == 'number') && 
					(msg.intens >= 0) && 
					(msg.intens <= 15) && 
					(msg.intens != node.intens)) {
					sysmodule.intensity(node.spich, msg.intens);
					node.intens = msg.intens;
				}
				return;
			}

			var npos = node.npos;

			if (msg.hasOwnProperty('npos')) {
				if ((typeof msg.npos == 'number') &&  (msg.npos < node.nmod))
					npos = msg.npos;
				else {
					syslib.outError(node, "inv position", "invalid postion or pos >= number module");
					return;
				}
			}

			// Load custom font
			if (msg.hasOwnProperty('customfont')) {
				var topos = 0;
				if (msg.hasOwnProperty('topos')) {
					if ((typeof msg.topos == 'number') &&  (msg.topos >= 0x00) && (msg.topos <= 0x1F))
						topos = msg.topos;
				}

				if (!Array.isArray(msg.customfont)) {
					syslib.outError(node, "inv font", "invalid custom font array");
					return;
				}
				
				for(var i=0; i < msg.customfont.length; i++) {
					if (typeof msg.customfont[i] !== "number") {
						syslib.outError(node, "font number", "invalid custom font number");
						return;
					}
	
					if (msg.customfont[i] < 0)
						msg.customfont[i] = 0;
					else if (msg.customfont[i] > 255)
						msg.customfont[i] = 255;
				}

				if (!sysmodule.write7segCustomFont(msg.customfont, topos)) {
					syslib.outError(node, "font error", "custom Font not load");
					return;
				}
				else
					syslib.outOk(node);
			}

			var ralign = node.ralign;

			if (msg.hasOwnProperty('ralign')) {
				if (typeof msg.ralign == 'boolean')
					ralign = msg.ralign;
			}

			var fill = node.fill;

			if (msg.hasOwnProperty('fill')) {
				if ((typeof msg.fill == 'string') && (msg.fill.length > 0))
					fill = msg.fill;
			}

			var out_txt = String(msg.payload);
			if (typeof  out_txt !== "string") {
				syslib.outError(node, "not string", "msg.payload not string");
				return;
			}

			if (out_txt.length == 0)
				return;

			if (!sysmodule.write7seg(node.spich, npos, ralign, fill, out_txt))
				syslib.outError(node, "write error", "device not write, check spi and device");
			else
				syslib.outOk(node);
		});

		node.on('close', function () {
			sysmodule.clear(node.spich);
		});
	});
}


max7219-mtx.js
/**
 * Copyright 2022 Ocean (iiot2k@gmail.com).
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 **/

"use strict";

module.exports = function(RED) {
	const syslib = require("./lib/syslib.js");
	const sysmodule = syslib.LoadModule("rpi_max7219");

    RED.nodes.registerType("max7219-mtx", function(n) {
		var node = this;
		RED.nodes.createNode(node, n);

		node.spich = parseInt(n.spich);
		node.nfont = parseInt(n.nfont);
		node.nmod = n.nmod;
		node.npos = n.npos;
		node.intens = n.intens;
		node.iserror = false;
		node.status({});
		node.name = this.name || "matrix-led " + node.spich + "#" + node.npos + "/" + node.nmod;

		if (sysmodule === undefined)
			node.iserror = syslib.outError(node, "driver error", "driver not load, wrong os or not Raspi");
		else if (node.npos >= n.nmod)
			node.iserror = syslib.outError(node, "inv position", "invalid position or >= number module");
		else if (!sysmodule.init(node.spich, node.nmod, node.intens))
			node.iserror = syslib.outError(node, "open error", "spi port not open, check spi settings");

		node.on("input", function (msg) {
			if (node.iserror)
				return;
			
			// Clear all leds	
			if (msg.hasOwnProperty('clear'))
				sysmodule.clear(node.spich);

			// Set led intensity
			if (msg.hasOwnProperty('intens')) {
				if ((typeof msg.intens == 'number') && 
					(msg.intens >= 0) && 
					(msg.intens <= 15) && 
					(msg.intens != node.intens)) {
					sysmodule.intensity(node.spich, msg.intens);
					node.intens = msg.intens;
				}
				return;
			}

			// Select postion
			var npos = node.npos;

			if (msg.hasOwnProperty('npos')) {
				if ((typeof msg.npos == 'number') &&  (msg.npos < node.nmod))
					npos = msg.npos;
				else {
					syslib.outError(node, "inv position", "invalid postion or pos >= number module");
					return;
				}
			}

			// Display pattern array
			if (Array.isArray(msg.payload)) {
				if (msg.payload.length == 0)
					return;

				for(var i=0; i < msg.payload.length; i++) {
					if (typeof msg.payload[i] !== "number") {
						syslib.outError(node, "not number", "msg.payload not array number");
						return;
					}
	
					if (msg.payload[i] < 0)
						msg.payload[i] = 0;
					else if (msg.payload[i] > 255)
						msg.payload[i] = 255;
				}
				
				if (!sysmodule.writeMatrix(node.spich, npos, msg.payload))
					syslib.outError(node, "write error", "device not write, check spi and device");
				else
					syslib.outOk(node);

				return;
			}

			// Select font
			var nfont = node.nfont;

			if (msg.hasOwnProperty('nfont')) {
				if ((typeof msg.nfont == 'number') &&  (msg.nfont < sysmodule.getFontCount()))
					nfont = msg.nfont;
				else {
					syslib.outError(node, "inv font", "invalid font number");
					return;
				}
			}

			// Load custom font
			if (msg.hasOwnProperty('customfont')) {
				var topos = 0;
				if (msg.hasOwnProperty('topos')) {
					if ((typeof msg.topos == 'number') &&  (msg.topos >= 0x00) && (msg.topos <= 0x1F))
						topos = msg.topos;
				}

				if (!Array.isArray(msg.customfont)) {
					syslib.outError(node, "inv font", "invalid custom font array");
					return;
				}
				
				for(var i=0; i < msg.customfont.length; i++) {
					if (typeof msg.customfont[i] !== "number") {
						syslib.outError(node, "font number", "invalid custom font number");
						return;
					}
	
					if (msg.customfont[i] < 0)
						msg.customfont[i] = 0;
					else if (msg.customfont[i] > 255)
						msg.customfont[i] = 255;
				}

				if (!sysmodule.writeCustomFont(msg.customfont, topos)) {
					syslib.outError(node, "font error", "custom Font not load");
					return;
				}
				else
					syslib.outOk(node);
			}

			// Display string/number
			var out_txt = String(msg.payload);
			if (typeof  out_txt !== "string") {
				syslib.outError(node, "not string", "msg.payload not string");
				return;
			}

			if (out_txt.length == 0)
				return;
	
			if (!sysmodule.writeMatrixTxt(node.spich, npos, nfont, out_txt))
				syslib.outError(node, "write error", "device not write, check spi and device");
			else
				syslib.outOk(node);
		});

		node.on('close', function () {
			sysmodule.clear(node.spich);
		});
	});
}
